File: src/Wc3Handle/Trigger.ts
import { Action, ActionList, Log } from "../Wc3Utils/index";
import { Handle } from "./Handle";

export class Trigger extends Handle<jtrigger> {
    constructor(){
        super(CreateTrigger())
        TriggerAddAction(this.handle, Trigger.runActions)
    }
    public static get(id: jtrigger | number){
        let instance = Handle.get(id)
        if (!instance){return}
        if (Handle.getWc3Type(instance.handle) != 'trigger'){
            Log.err('Trigger: got wrong type of handle.', 2)
        }
        return instance as Trigger
    }
    public static getTriggering(){return Trigger.get(GetTriggeringTrigger())}

    public get handle(){return this._handle as jtrigger}

    public addAction(callback: (this: void, trig: Trigger)=>void){
        return this._actions.add(callback)
    }

    public removeAction(action: Action<[Trigger], void>){
        return this._actions.remove(action)
    }

    private static runActions(this: void){
        let trig = Trigger.getTriggering()
        trig?._actions.run(trig)
    }

    public destroy(){
        DestroyTrigger(this.handle)
        super.destroy()
    }

    private _actions = new ActionList<[Trigger]>()
}

File: src/Wc3HandleLua/Handle/Trigger.lua
--=========
-- Include
--=========

local Class = LibManager.getDepency('LuaClass') or error('')
---@type Wc3Utils
local Wc3Utils = LibManager.getDepency('Wc3Utils') or error('')
local ActionList = Wc3Utils.ActionList or error('')
local isTypeErr = Wc3Utils.isTypeErr or error('')

---@type HandleClass
local Handle = require('Handle') or error('')

--=======
-- Class
--=======

local Trigger = Class.new('Trigger', Handle, ActionList)
---@class Trigger : Handle, ActionList
---@field add fun(self:Trigger, callback:Callback):Action
---@field remove fun(self:Trigger, action:Action):boolean
---@field get fun(self:Trigger, pos:number):Action | nil
---@field count fun(self:Trigger):number
---@field clear fun(self:Trigger)
---@field run fun(self:Trigger)
local public = Trigger.public
---@class TriggerClass : HandleClass, ActionListClass
local static = Trigger.static
---@type TriggerClass
local override = Trigger.override
local private = {}

--========
-- Static
--========

---@param child Trigger | nil
---@return Trigger
function override.new(child)
    if child then isTypeErr(child, Trigger, 'child') end

    local instance = child or Class.allocate(Trigger)
    instance = ActionList.new(nil, instance)
    instance = Handle.new(CreateTrigger(), DestroyTrigger, instance)
    TriggerAddAction(instance:getData(), private.runEventActions)

    return instance
end

--========
-- Public
--========

--=========
-- Private
--=========

private.data = setmetatable({}, {__mode = 'k'})

function private.runEventActions()
    local self = static.getLinked(GetTriggeringTrigger())
    self:run()
end

return static